core/chains/evm/gas/rollups/l1_oracle.go

```diff
@@ -0,0 +1,313 @@
+		gasPriceMethod = ArbGasInfo_getL1BaseFeeEstimate
+		l1GasPriceMethodAbi, gasPriceErr = abi.JSON(strings.NewReader(GetL1BaseFeeEstimateAbiString))
+		l1GasCostAddress = ArbNodeInterfaceAddress
+		gasCostMethod = ArbNodeInterface_gasEstimateL1Component
+		l1GasCostMethodAbi, gasCostErr = abi.JSON(strings.NewReader(GasEstimateL1ComponentAbiString))
+	case config.ChainOptimismBedrock:
+		l1GasPriceAddress = OPGasOracleAddress
+		gasPriceMethod = OPGasOracle_l1BaseFee
+		l1GasPriceMethodAbi, gasPriceErr = abi.JSON(strings.NewReader(L1BaseFeeAbiString))
+		l1GasCostAddress = OPGasOracleAddress
+		gasCostMethod = OPGasOracle_getL1Fee
+		l1GasCostMethodAbi, gasCostErr = abi.JSON(strings.NewReader(GetL1FeeAbiString))
+	case config.ChainKroma:
+		l1GasPriceAddress = KromaGasOracleAddress
+		gasPriceMethod = KromaGasOracle_l1BaseFee
+		l1GasPriceMethodAbi, gasPriceErr = abi.JSON(strings.NewReader(L1BaseFeeAbiString))
+		l1GasCostAddress = ""
+		gasCostMethod = ""
+	case config.ChainScroll:
+		l1GasPriceAddress = ScrollGasOracleAddress
+		gasPriceMethod = ScrollGasOracle_l1BaseFee
+		l1GasPriceMethodAbi, gasPriceErr = abi.JSON(strings.NewReader(L1BaseFeeAbiString))
+		l1GasCostAddress = ScrollGasOracleAddress
+		gasCostMethod = ScrollGasOracle_getL1Fee
+		l1GasCostMethodAbi, gasCostErr = abi.JSON(strings.NewReader(GetL1FeeAbiString))
+	default:
+		panic(fmt.Sprintf("Received unspported chaintype %s", chainType))
+	}
+
+	if gasPriceErr != nil {
+		panic(fmt.Sprintf("Failed to parse L1 gas price method ABI for chain: %s", chainType))
+	}
+	if gasCostErr != nil {
+		panic(fmt.Sprintf("Failed to parse L1 gas cost method ABI for chain: %s", chainType))
+	}
+
+	return &l1Oracle{
+		client:     ethClient,
+		pollPeriod: PollPeriod,
+		logger:     logger.Sugared(logger.Named(lggr, fmt.Sprintf("L1GasOracle(%s)", chainType))),
+		chainType:  chainType,
+
+		l1GasPriceAddress:   l1GasPriceAddress,
+		gasPriceMethod:      gasPriceMethod,
+		l1GasPriceMethodAbi: l1GasPriceMethodAbi,
+		l1GasCostAddress:    l1GasCostAddress,
+		gasCostMethod:       gasCostMethod,
+		l1GasCostMethodAbi:  l1GasCostMethodAbi,
+
+		chInitialised: make(chan struct{}),
+		chStop:        make(chan struct{}),
+		chDone:        make(chan struct{}),
+	}
+}
+
+func (o *l1Oracle) Name() string {
+	return o.logger.Name()
+}
+
+func (o *l1Oracle) Start(ctx context.Context) error {
+	return o.StartOnce(o.Name(), func() error {
+		go o.run()
+		<-o.chInitialised
+		return nil
+	})
+}
+func (o *l1Oracle) Close() error {
+	return o.StopOnce(o.Name(), func() error {
+		close(o.chStop)
+		<-o.chDone
+		return nil
+	})
+}
+
+func (o *l1Oracle) HealthReport() map[string]error {
+	return map[string]error{o.Name(): o.Healthy()}
+}
+
+func (o *l1Oracle) run() {
+	defer close(o.chDone)
+
+	t := o.refresh()
+	close(o.chInitialised)
+
+	for {
+		select {
+		case <-o.chStop:
+			return
+		case <-t.C:
+			t = o.refresh()
+		}
+	}
+}
+
+func (o *l1Oracle) refresh() (t *time.Timer) {
```

Comment:
Maybe for a future PR.
But we ought to have some logic to start giving errors to all methods from the l1Oracle, if we fail to fetch fresh prices for last 12-20 seconds.
Otherwise with current logic, we keep using stale prices, which may be worse behavior for applications.
Perhaps if we just report errors in such cases, it may be better.

Reply:
I added tracking for when the price was refreshed so we can validate for stale prices on read. Is this what you had in mind?

Reply:
Yes, perfect!